cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -529,47 +541,118 @@ public Void call() throws Exception {
+
+    // multiple input formats configured
+    for (Map.Entry<String, MapperInput> mapperInputEntry : mapperInputs.entrySet()) {
+      MapperInput mapperInput = mapperInputEntry.getValue();
+      InputFormatProvider provider = mapperInput.getInputFormatProvider();
+      Map<String, String> inputFormatConfiguration = new HashMap<>(provider.getInputFormatConfiguration());
+
+      String inputName = mapperInputEntry.getKey();
+      Class<? extends Mapper> mapperClass = mapperInput.getMapper();
 
       // A bit hacky for stream.
-      // For stream, we need to do two extra steps.
-      // 1. stream usage registration since it only happens on client side.
-      // 2. Infer the stream event decoder from Mapper/Reducer
       if (provider instanceof StreamInputFormatProvider) {
-        StreamInputFormatProvider streamProvider = (StreamInputFormatProvider) provider;
-        Type inputValueType = getInputValueType(jobConf, StreamEvent.class);
-        ConfigurationUtil.setAll(streamProvider.setDecoderType(new HashMap<String, String>(), inputValueType), jobConf);
-
-        Id.Stream streamId = streamProvider.getStreamId();
-        try {
-          usageRegistry.register(context.getProgram().getId(), streamId);
-          streamAdmin.addAccess(new Id.Run(context.getProgram().getId(), context.getRunId().getId()),
-                                streamId, AccessType.READ);
-        } catch (Exception e) {
-          LOG.warn("Failed to register usage {} -> {}", context.getProgram().getId(), streamId, e);
-        }
+        setDecoderForStream((StreamInputFormatProvider) provider, job, inputFormatConfiguration,
+                            mapperClass == null ? rootMapperTypeToken : resolveClass(mapperClass, Mapper.class));
       }
+
+      MultipleInputs.addInput(job, inputName,
+                              provider.getInputFormatClassName(), inputFormatConfiguration, mapperClass);
+    }
+  }
+
+  private void setDecoderForStream(StreamInputFormatProvider streamProvider, Job job,
+                                   Map<String, String> inputFormatConfiguration, TypeToken<?> mapperTypeToken) {
+    // For stream, we need to do two extra steps.
+    // 1. stream usage registration since it only happens on client side.
+    // 2. Infer the stream event decoder from Mapper/Reducer
+    Type inputValueType = getInputValueType(job.getConfiguration(), StreamEvent.class, mapperTypeToken);
+    streamProvider.setDecoderType(inputFormatConfiguration, inputValueType);
+
+    Id.Stream streamId = streamProvider.getStreamId();
+    try {
+      usageRegistry.register(context.getProgram().getId(), streamId);
+      streamAdmin.addAccess(new Id.Run(context.getProgram().getId(), context.getRunId().getId()),
+                            streamId, AccessType.READ);
+    } catch (Exception e) {
+      LOG.warn("Failed to register usage {} -> {}", context.getProgram().getId(), streamId, e);
```

Comment:
I split it up into two separate try-catch blocks with separate messages, but that looks messy. I think its fine to have it like this, because the logged exception message and stack trace will give the answer to that question.